lib/core/src/Cardano/Pool/Metadata.hs

```diff
@@ -65,23 +95,123 @@ type Api
 -- | A client for fetching metadata from an Aggregation server.
 --
 -- See also 'newClient' to construct a client.
-newtype Client api = Client
+newtype Client api m = Client
     { getStakePoolMetadata
         :: PoolId
-        -> IO (Either ClientError StakePoolOffChainMetadata)
+        -> m (Either ClientError (Maybe StakePoolOffChainMetadata))
     }
 
-mkClient
+mkClientIO
     :: Manager
     -> BaseUrl
-    -> Client Api
-mkClient mgr baseUrl = Client
-    { getStakePoolMetadata = \pid ->
-        fmap getApiT <$> run (getMetadata (ApiT pid))
+    -> Client Api IO
+mkClientIO mgr baseUrl = Client
+    { getStakePoolMetadata = \pid -> do
+        run (getMetadata (ApiT pid)) <&> \case
+            Right (ApiT meta) ->
+                Right (Just meta)
+
+            Left (FailureResponse _ res) | responseStatusCode res == status404 -> do
+                Right Nothing
+
+            Left e ->
+                Left e
     }
   where
     run :: ClientM a -> IO (Either ClientError a)
     run query = runClientM query (mkClientEnv mgr baseUrl)
 
     getMetadata =
         client (Proxy @Api)
+
+-- | A configuration for managing metadata with the aggregation server.
+-- Callbacks and parameterized effects allows for easier testing while a real
+-- specialization would wire a database connector in here.
+data MetadataConfig (m :: * -> *) = MetadataConfig
+    { saveMetadata
+        :: PoolId -> (Maybe StakePoolOffChainMetadata, UTCTime) -> m ()
+        -- ^ A callback action for storing an off-chain metadata. The callback
+        -- may be called with 'Nothing' to store that no metadata were found for
```

Comment:
> may be called with 'Nothing' to store that no metadata were found

Does this mean the metadata will be removed?

> this allows for not constantly re-fetching data for pools that are known to have no metadata

Not quite understanding this from the comment

Reply:
> Does this mean any existing metadata will be removed?

No. This would be done automatically by the DB when pushing a de-registration certificate.

> Not quite understanding this from the comment

The thing is: we don't want to only cache metadata when they exist, but also when they aren't available. If we only cache metadata hit, then everytime we'll "refresh" metadata we'll have to make a query for all the pools that didn't have metadata. 

Yet, if we cache both the presence or absence of metadata, we only need a request for pools that are new and for which the state is unknown.